staging/src/k8s.io/apiextensions-apiserver/pkg/storageversion/manager.go

```diff
@@ -0,0 +1,268 @@
+	genericstorageversion "k8s.io/apiserver/pkg/storageversion"
+	svInformers "k8s.io/client-go/informers/apiserverinternal/v1alpha1"
+)
+
+var (
+	// workerFrequency is the frequency at which the worker for updating storageversion for a CRD
+	// will be called.
+	workerFrequency = 1 * time.Second
+)
+
+// Manager maintains necessary structures needed for updating
+// StorageVersion for CRDs. It does goroutine management to allow CRD
+// storage version updates running in the background and not blocking the caller.
+type Manager struct {
+	// client is the client interface that manager uses to update
+	// StorageVersion objects.
+	client genericstorageversion.Client
+	// apiserverID is the ID of the apiserver that invokes this manager.
+	apiserverID string
+	crdInformer crdinformers.CustomResourceDefinitionInformer
+	svInformer  svInformers.StorageVersionInformer
+	// workqueue to process storageversion updates for this CRD.
+	queue *workqueue.Type
+	// map from crd.Name -> number of active teardowns for this CRD.
+	storageVersionUpdateInfoMap sync.Map
+}
+
+type storageVersionUpdateInfo struct {
+	crd *apiextensionsv1.CustomResourceDefinition
+	// processedCh is closed by the storage version manager after the
+	// storage version update gets processed successfully.
+	// The API server will unblock and allow CR write requests if this
+	// channel is closed.
+	processedCh chan struct{}
+
+	// teardownFinishedCh is closed when the teardown of a previous
+	// storageversion completes. This is used by the storageversion manager
+	// to unblock publishing of the latest storageversion.
+	teardownCount int
+}
+
+// NewManager creates a CRD StorageVersion Manager.
+func NewManager(svClient genericstorageversion.Client, apiserverID string,
+	crdInformer crdinformers.CustomResourceDefinitionInformer, svInformer svInformers.StorageVersionInformer) *Manager {
+	return &Manager{
+		client:                      svClient,
+		apiserverID:                 apiserverID,
+		crdInformer:                 crdInformer,
+		svInformer:                  svInformer,
+		queue:                       workqueue.NewNamed("storageversion-updater"),
+		storageVersionUpdateInfoMap: sync.Map{},
+	}
+}
+
+// UpdateActiveTeardownsCount updates the teardown count of the CRD in sync.Map.
+func (m *Manager) UpdateActiveTeardownsCount(crdName string, value int) error {
+	val, ok := m.storageVersionUpdateInfoMap.Load(crdName)
+	if !ok {
+		return fmt.Errorf("error while trying to update number of teardowns. No entry found in sync.Map for %v", crdName)
+	}
+	svInfo := val.(*storageVersionUpdateInfo)
+	svInfo.teardownCount += value
+	return nil
+}
+
+func (m *Manager) UpdateSVUpdateInfoAndEnqueue(crd *apiextensionsv1.CustomResourceDefinition, processedCh chan struct{}) {
+	val, _ := m.storageVersionUpdateInfoMap.LoadOrStore(crd.Name, &storageVersionUpdateInfo{})
+	svInfo := val.(*storageVersionUpdateInfo)
+	svInfo.crd = crd
+	if processedCh != nil {
+		svInfo.processedCh = processedCh
```

Comment:
Can we use the `Store` method to update the `sync.Map`, to make sure the update is atomic?

Reply:
Done.